design-documents/graph-ql/coverage/storefront-route.md

```diff
@@ -0,0 +1,292 @@
+# Proposal for Representing Storefront Routes in GraphQL
+
+## Problem Statement
+
+**URL routing in the PWA must handle all custom URLs maintained by the merchant, installed by extensions, and built in to Magento.**
+
+PWAs have client-side routing logic so they can work offline.
+The Magento PWA includes a [router implementation](https://reacttraining.com/react-router/web/guides/quick-start) which receives dispatched URLs and handles them.
+Some of these routes are hardcoded into the PWA logic, but most routes are created and controlled by the business user.
+The business user can modify the custom URL of any entity at any time.
+Therefore, the PWA must query the Magento API to learn how to route a URL and display its associated entity.
+
+This document proposes a GraphQL-driven solution to the routing issue that simplifies the routing scheme, supports business users, and enables extensibility.
+
+## Requirements
+
+As stated above, the PWA must be able to route:
+
+1) SEO URLs for catalog and CMS pages
+2) Built-in Magento routes, like `/checkout`
+3) Custom routes installed by extensions and other modules
+
+The current PWA implementation has a partial solution for **(1)**, using the [`urlResolver`](https://github.com/magento/graphql-ce/issues/80).
+It has a suboptimal solution for **(2)**, in the form of reimplementing Magento routing logic in the client-side app router.
+There is no current solution for **(3)**.
+
+To support Magento's SEO and display customization features, the PWA must be able to query the API for the following data:
+
+- **Entity type**, page type or content type (e.g. category, product, or CMS page)
+- **Entity** content and data
+
+With that information, the PWA can asynchronously load the view components which handle page display for the appropriate entity.
+
+Additionally, the PWA should be able to query the API for **display metadata** for any entity.
+This could be a simple string representing a shared "variant" component label, which would enable PWAs to support the equivalent of per-entity Theme Variants in existing Magento 2.
+
+<!-- diagram: url => { entity_type, entity, metadata } -->
+
+The current `urlResolver` query delivers part of these requirements: entity type in the form of an enum of "magic strings", such as `CATEGORY` and `CMS_PAGE`, and entity data in the form of a unique ID, which functions as a foreign key to be used in a subsequent GraphQL call.
+
+<!-- diagram: url => { type, id } -->
+
+This solution helps us honor SEO URLs, which is a _distinguishing feature_ of PWA Studio.
+However, it also has flaws:
+
+- The entity type enum must be hand-maintained and mapped to real types
+- Not all routable entities have implemented types: only `CATEGORY`, `CMS_PAGE` and `PRODUCT`.
```

Comment:
Do we want to support client-side redirects as page type? It is possible to define redirects in Magento, however they will not be honored by PWA as far as I understand.

Reply:
I see. I think it should be possible to implement a redirect. Each RootComponent owns the behavior of the app once the entity has loaded. If the entity is a redirect, then the RootComponent can simply update the URL and load the provided location.